metabase_permissions_group panics on create

When attempting to create the `metabase_permissions_group` resource, the provider panics:

```console
$ terraform apply

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # metabase_permissions_group.example will be created
  + resource "metabase_permissions_group" "example" {
      + id   = (known after apply)
      + name = "example"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

metabase_permissions_group.example: Creating...
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-metabase_v0.8.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x104b9b444]

goroutine 37 [running]:
github.com/flovouin/terraform-provider-metabase/internal/provider.(*PermissionsGroupResource).Create(0x14000412e70, {0x1051265a8, 0x14000512330}, {{{{0x10512b3f8, 0x14000512900}, {0x10500afa0, 0x14000512870}}, {0x10512d6c8, 0x1400042a960}}, {{{0x10512b3f8, ...}, ...}, ...}, ...}, ...)
        github.com/flovouin/terraform-provider-metabase/internal/provider/permissions_group_resource.go:81 +0x1f4
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).CreateResource(0x140002821e0, {0x1051265a8, 0x14000512330}, 0x140000c35c8, 0x140000c35a0)
        github.com/hashicorp/terraform-plugin-framework@v1.11.0/internal/fwserver/server_createresource.go:101 +0x400
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ApplyResourceChange(0x140002821e0, {0x1051265a8, 0x14000512330}, 0x1400042b2c0, 0x140000c36c0)
        github.com/hashicorp/terraform-plugin-framework@v1.11.0/internal/fwserver/server_applyresourcechange.go:57 +0x380
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).ApplyResourceChange(0x140002821e0, {0x1051265a8?, 0x14000512240?}, 0x1400042b220)
        github.com/hashicorp/terraform-plugin-framework@v1.11.0/internal/proto6server/server_applyresourcechange.go:55 +0x2e0
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0x14000294aa0, {0x1051265a8?, 0x14000475830?}, 0x14000434f50)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov6/tf6server/server.go:865 +0x2a8
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0x1050f5280, 0x14000294aa0}, {0x1051265a8, 0x14000475830}, 0x14000504380, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:518 +0x1c0
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000192e00, {0x1051265a8, 0x140004757a0}, {0x10512c380, 0x140000e6000}, 0x1400050a360, 0x1400038c270, 0x105662278, 0x0)
        google.golang.org/grpc@v1.66.0/server.go:1393 +0xb78
google.golang.org/grpc.(*Server).handleStream(0x14000192e00, {0x10512c380, 0x140000e6000}, 0x1400050a360)
        google.golang.org/grpc@v1.66.0/server.go:1804 +0xb20
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.66.0/server.go:1029 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 8
        google.golang.org/grpc@v1.66.0/server.go:1040 +0x13c

Error: The terraform-provider-metabase_v0.8.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

My config file looked like:

```hcl
terraform {
  required_providers {
    metabase = {
      source = "flovouin/metabase"
    }
  }
}

provider "metabase" {
  endpoint = "http://127.0.0.1:3000"

  api_key = "mb_Q5D0iER0AXSS8Zdr0YsQ73+Dzr74uRlYB0nmB0tSfAw="
}

resource "metabase_permissions_group" "example" {
  name = "example"
}
```

I tested against a fresh local instance of metabase from the official docker container.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+BUG FIXES:
+
+- Catch and return an error when the API returns a response that could not be parsed.
+
 ## 0.8.0 (2024-09-08)
 
 BREAKING CHANGES:

--- a/internal/provider/utils.go
+++ b/internal/provider/utils.go
@@ -72,6 +72,15 @@ func checkMetabaseResponse(r metabase.MetabaseResponse, err error, statusCodes [
 		}
 	}
 
+	if r.HasExpectedStatusWithoutExpectedBody() {
+		return diag.Diagnostics{
+			diag.NewErrorDiagnostic(
+				fmt.Sprintf("Unexpected response while calling the Metabase API for operation '%s'.", operation),
+				fmt.Sprintf("Status code: %d, failed to parse body: %s", r.StatusCode(), r.BodyString()),
+			),
+		}
+	}
+
 	for _, s := range statusCodes {
 		if r.StatusCode() == s {
 			return diag.Diagnostics{}

--- a/metabase/metabase_response.go
+++ b/metabase/metabase_response.go
@@ -5,120 +5,237 @@ package metabase
 type MetabaseResponse interface {
 	StatusCode() int
 	BodyString() string
+	HasExpectedStatusWithoutExpectedBody() bool
 }
 
 func (r *CreateCardResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *CreateCardResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *GetCardResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *GetCardResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *UpdateCardResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *UpdateCardResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *GetCollectionPermissionsGraphResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *GetCollectionPermissionsGraphResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *ReplaceCollectionPermissionsGraphResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *ReplaceCollectionPermissionsGraphResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *CreateCollectionResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *CreateCollectionResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *GetCollectionResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *GetCollectionResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *UpdateCollectionResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *UpdateCollectionResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *ListCollectionItemsResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *ListCollectionItemsResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *CreateDashboardResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *CreateDashboardResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *GetDashboardResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *GetDashboardResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *UpdateDashboardResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *UpdateDashboardResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *DeleteDashboardResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *DeleteDashboardResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return false
+}
+
 func (r *CreateDatabaseResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *CreateDatabaseResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *GetDatabaseResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *GetDatabaseResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *UpdateDatabaseResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *UpdateDatabaseResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *DeleteDatabaseResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *DeleteDatabaseResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return false
+}
+
 func (r *GetPermissionsGraphResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *GetPermissionsGraphResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *ReplacePermissionsGraphResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *ReplacePermissionsGraphResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *CreatePermissionsGroupResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *CreatePermissionsGroupResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *GetPermissionsGroupResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *GetPermissionsGroupResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *UpdatePermissionsGroupResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *UpdatePermissionsGroupResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *DeletePermissionsGroupResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *DeletePermissionsGroupResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return false
+}
+
 func (r *CreateSessionResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *CreateSessionResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *ListTablesResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *ListTablesResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *GetTableMetadataResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *GetTableMetadataResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }
 
 func (r *UpdateTableResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *UpdateTableResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *GetFieldResponse) BodyString() string {
 	return string(r.Body)
 }
 
+func (r *GetFieldResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
+}
+
 func (r *UpdateFieldResponse) BodyString() string {
 	return string(r.Body)
+}
+
+func (r *UpdateFieldResponse) HasExpectedStatusWithoutExpectedBody() bool {
+	return r.StatusCode() == 200 && r.JSON200 == nil
 }

